src/app.rs

```diff
@@ -47,6 +47,138 @@ pub struct App {
     display_help: bool,
     pub is_searching: bool,
     pub channel_text_width: usize,
+    receipt_handler: ReceiptHandler,
+}
+
+#[derive(Debug, Default, PartialEq, Eq)]
+pub struct ReceiptHandler {
+    receipt_set: HashMap<Uuid, ReceiptQueues>,
+    time_since_update: u64,
+}
+
+impl ReceiptHandler {
+    pub fn new() -> Self {
+        Self {
+            receipt_set: HashMap::new(),
+            time_since_update: 0u64,
+        }
+    }
+
+    pub fn add_receipt_event(&mut self, event: ReceiptEvent) {
+        // Add a new set in the case no receipt had been handled for this contact
+        // over the current session
+        self.receipt_set
+            .entry(event.uuid)
+            .or_insert_with(ReceiptQueues::new);
+
+        self.receipt_set
+            .get_mut(&event.uuid)
+            .unwrap()
+            .add(event.timestamp, event.receipt_type);
+    }
+
+    // Dictates whether receipts should be sent on the current tick
+    // Not used for now as
+    fn do_tick(&mut self) -> bool {
+        true
+    }
+
+    pub fn step(&mut self, signal_manager: &dyn SignalManager) -> bool {
+        if !self.do_tick() {
+            return false;
+        }
+        if self.receipt_set.is_empty() {
+            return false;
+        }
+        let uuid = *self.receipt_set.keys().next().unwrap();
+        let mut queue = self.receipt_set.remove(&uuid).unwrap();
```

Comment:
You can use `.remove_entry()` to get id and queue at the same time. I think however, it is better to use the `.entry` API to get an entry, and if it is not vacant, to do the logic below. And in-case, its value, the queue, is empty, to remove it.